File: backend/app/models/yearly_goal.py
import enum
from datetime import datetime
from sqlalchemy import Column, String, Numeric, Integer, Date, Enum, ForeignKey, Text, DateTime
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import uuid

from app.db.base import Base


class GoalStatus(str, enum.Enum):
    PENDING = "pending"
    IN_PROGRESS = "in-progress"
    COMPLETED = "completed"
    PAUSED = "paused"


class GoalCategory(str, enum.Enum):
    HEALTH = "health"
    CAREER = "career"
    LEARNING = "learning"
    FINANCE = "finance"
    RELATIONSHIP = "relationship"
    ENTERTAINMENT = "entertainment"


class YearlyGoal(Base):
    __tablename__ = "yearly_goals"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    year = Column(Integer, nullable=False)
    title = Column(String(200), nullable=False)
    description = Column(Text)
    category = Column(Enum(GoalCategory, values_callable=lambda x: [e.value for e in x]), nullable=False)
    color = Column(String(7))
    target_value = Column(Numeric(10, 2), nullable=False)
    current_value = Column(Numeric(10, 2), default=0)
    unit = Column(String(20))
    status = Column(Enum(GoalStatus, values_callable=lambda x: [e.value for e in x]), default=GoalStatus.PENDING)
    start_date = Column(Date)
    end_date = Column(Date)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    user = relationship("User", back_populates="yearly_goals")
    monthly_milestones = relationship("MonthlyMilestone", back_populates="yearly_goal", cascade="all, delete-orphan")
    # TODO: Add monthly_plans relationship when MonthlyPlan model is implemented
    # monthly_plans = relationship("MonthlyPlan", back_populates="yearly_goal", cascade="all, delete-orphan")

    @property
    def completion_rate(self) -> float:
        if self.target_value and float(self.target_value) > 0:
            return round(float(self.current_value) / float(self.target_value) * 100, 2)
        return 0.0

    @property
    def milestones(self):
        return self.monthly_milestones

    def __repr__(self):
        return f"<YearlyGoal {self.year}: {self.title}>"


class MonthlyMilestone(Base):
    __tablename__ = "monthly_milestones"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    yearly_goal_id = Column(UUID(as_uuid=True), ForeignKey("yearly_goals.id", ondelete="CASCADE"), nullable=False)
    month = Column(Integer, nullable=False)
    target_value = Column(Numeric(10, 2), nullable=False)
    achieved_value = Column(Numeric(10, 2), default=0)
    note = Column(Text)

    # Relationships
    yearly_goal = relationship("YearlyGoal", back_populates="monthly_milestones")

    def __repr__(self):
        return f"<MonthlyMilestone month={self.month} target={self.target_value}>"

File: frontend/src/types/yearlyGoal.ts
export type GoalCategory =
  | "health"
  | "career"
  | "learning"
  | "finance"
  | "relationship"
  | "entertainment";

export type GoalStatus = "pending" | "in-progress" | "completed" | "paused";

export interface MonthlyMilestone {
  id?: string;
  month: number;
  target_value: number;
  achieved_value: number;
  note?: string;
}

export interface YearlyGoal {
  id: string;
  user_id: string;
  year: number;
  title: string;
  description?: string;
  category: GoalCategory;
  color: string;
  target_value: number;
  current_value: number;
  unit?: string;
  status: GoalStatus;
  start_date?: string;
  end_date?: string;
  completion_rate: number;
  milestones: MonthlyMilestone[];
  created_at: string;
  updated_at: string;
}

export interface YearlyGoalCreate {
  year: number;
  title: string;
  description?: string;
  category: GoalCategory;
  color?: string;
  target_value: number;
  unit?: string;
  start_date?: string;
  end_date?: string;
  auto_generate_milestones?: boolean;
}

export interface YearlyGoalUpdate {
  title?: string;
  description?: string;
  color?: string;
  target_value?: number;
  status?: GoalStatus;
  end_date?: string;
}

export interface ProgressUpdate {
  progress: number;
  month?: number;
  note?: string;
}

export const GOAL_CATEGORIES: { value: GoalCategory; label: string; color: string }[] = [
  { value: "health", label: "健康", color: "#10B981" },
  { value: "career", label: "事业", color: "#3B82F6" },
  { value: "learning", label: "学习", color: "#8B5CF6" },
  { value: "finance", label: "财务", color: "#F59E0B" },
  { value: "relationship", label: "人际关系", color: "#EC4899" },
  { value: "entertainment", label: "娱乐", color: "#06B6D4" },
];

export const GOAL_STATUS: { value: GoalStatus; label: string; color: string }[] = [
  { value: "pending", label: "未开始", color: "#9CA3AF" },
  { value: "in-progress", label: "进行中", color: "#3B82F6" },
  { value: "completed", label: "已完成", color: "#10B981" },
  { value: "paused", label: "已暂停", color: "#F59E0B" },
];
